src/gui_common/CustomPopupMenu.cs

```diff
@@ -46,9 +45,43 @@ protected virtual void ResolveNodeReferences()
     {
     }
 
+    /// <summary>
+    ///   Calculates the size the pop should be taking into account minimum
+    ///   height needed by content
+    /// </summary>
+    /// <returns>Returns what should be the size of popup considering content height</returns>
+    protected virtual Vector2 CalculateSize()
+    {
+        RectSize = cachedMinSize;
+
+        var clipControl = GetNodeOrNull<Control>("Panel/Control");
```

Comment:
Why this is the "clip" control? This isn't clear at all to me what's going on here and why this uses a nodepath here to get some random child node.

Reply:
It is a node who, in spite of the un-descriptive default name, appears to have sole purpose of clipping the content of the container node and has a margin applied. Said margin is needed to properly size the height of the CustomPopupMenu. I'll fully admit that getting it with the node path was just something to get this functional at the time. I'll make it added via export ala the container and panel nodes if that is desired.

Reply:
That would be better, yes. If the clip control is entirely meant to clip the content, then maybe that is the root cause why the popup doesn't size itself correctly automatically?

Reply:
Thinking about this now that you mentioned it. It probably is the core crux of this issue and removing it probably will fix things in an ever so simpler way. Like, this clip control was probably implemented early on as a conscious design decision to have these pop-ups clip content rather than ever expand. Will experiment tomorrow. If it does fix this to remove this node then I'll probably revert prior commits and only change this then.

Reply:
If I remember right, it's probably me who added that special clipping control. I think the reason was purely to make it possible to animate the menu expanding and closing, otherwise the contents of the context menu holds the container tightly making it impossible to shrink completely. I guess this height not adjusting problem comes as an oversight.

Reply:
Apparently removing the node that clips content is not the end of the issue. Parent control nodes do not expand or grow to fit size of children. Overall dynamic sizing of controls isn't easily possible in Godot apparently outside of containers and their logic, hence the workarounds done before.

Reply:
I was thinking that *why* we aren't using a container here... Would it be possible to switch this to use a container to ensure the sizing is automatically correct? That would basically solve all the problems of trying to fight with the way the Godot GUI system works here. Though I don't know what's the original reason why a container was not used here.

Reply:
My off the tip of the head guess at why containers weren't used has to do with the fact that a panel is also used. As far as I can remember and know there is no default container that works well in having a panel as a background for content more or less. Will check to see if there actually is one however. Tho, in case there isn't one that works in that situation I do wonder if we can just have the panel as a sibling to the vbox and have the root control node of CustomPopupMenu more or less act as a container changing the panel to match vbox size. Or maybe just make a generic container that does similar logic. Mm, not sure if, or how that will work. Will have to look into it probably.

Reply:
I'm not sure that would work. Changing size or position of the panel as child of a container could maybe cause active fighting against the container logic I think. Like, if we up the size it may up the size of the container which could be bad because that maybe will affect how the logic will treat the content such as buttons and labels which may not be what we want. If position is changed, I'm not sure how that'll interact with the container either. Like will it consider the space where it thinks the panel is when calculating where the content should be and cause issues and leave empty space there?  Maybe these are just unnecessary concerns; won't know until an implementation is tried.

Reply:
If the container doesn't automatically resize the background to fit, then it's just a simple case of manually positioning the background node. I've used that approach before where I had controls that weren't sized and positioned by their parent (this was fully intentional as these places required special considerations). Godot Controls that don't get managed by a parent container can be manually moved without problems. For example this approach is used to overlap two controls on the same visual space when needed. Setting the correct mouse filters makes the cursor click the right elements. Even if the container does set the child sizing, it is possible to override that after it is set by registering the right callbacks, AFAIK this also doesn't leave to fighting as containers only reposition their children when necessary i.e. they don't check each frame that things are right.

Reply:
Mm, good news. There is apparently a container we could use, and it would actually remove the need to have a panel node at all. PanelContainer. It draws the panel behind the content with a stylebox. Should be functionally ideal if I read the docs correctly. The only thing that I imagine would be annoying about switching the nodes is that modifying inherited scenes in Godot is an utter trifle and not well supported in the editor if I remember correctly.

Reply:
Doesn't it just work by switching the node type in the scene? I doubt the change would need to be made in more than 2-3 files (.tscn and the C# file to specify the new base class). I've switched scene root node types before and it's not been a problem really.

Reply:
Not exactly. You'd also have to remove the panel node since that is no longer need with PanelContainer which may present problems with inherited scenes. Inherited scenes are... well not easy to work with. Like changing the type of the root node can sort of work; you just have to change it in every instance since that afaik doesn't update when the base scene gets changed. If you change the hierarchy of nodes, like removing the panel and clip making container child of root, stuff will not update correctly in the inherited scene if done normally from editor and you have to dig into the tscn file and manually mess with stuff instead which can be a hassle.

Overall inherited scenes are like making the children of a scene instance editable. If you seriously need it for stuff to function, then go ahead as a last resort, but if you're able to avoid it then alternatives are probably preferred 'cause otherwise stuff may be hard to maintain in long run when things change.

Tho maybe I've heard and learned wrong. Generally I've heard that composition is preferred over inheritance when dealing with Godot nodes if possible.

Reply:
Unless I misremember, we have just 2-3 inherited scenes from the popup. Everything else is just scene instances where I don't think many places would have editable child nodes enabled. So I don't foresee that this change would be that much work.

Reply:
It'll be also much easier to fix and do properly now rather than in 3 years when we have much more game done.

Reply:
Well, I've made progress. I'm fairly certain that changing the panel node to a PanelContainer and removing the clip control node should do intended behavior. Can't test however because somehow, some way, some change I've made in the process of this is now causing LineChart.cs lines 690 and 831 to give a null reference exception which is unexpected completely. Not sure how this is even affecting LineChart given it doesn't even use a CustomPopupMenu for it's popup or anything like that.

Reply:
...and now reloading the editor it's fixed itself? I honestly have no clue, but it is lightly concerning that a random null reference appeared and disappeared without logical reason.

Reply:
So, as far as I know this is the last of the outstanding changes requested so far. Is there any changes you want made or bugs found that need to be fixed?

Reply:
I'll read through the PR changes again today. I hopefully won't spot anything so that this can be merged (as it looks like there are no merge conflicts with master even with my huge refactor being merged).